src/Simulation/QCTraceSimulator/DepthCounter.cs

```diff
@@ -105,33 +116,57 @@ public void OnOperationStart(HashedString name, OperationFunctor functorSpeciali
             opRec.FunctorSpecialization = functorSpecialization;
             opRec.OperationName = name;
             opRec.InputQubitMetrics = Utils.UnboxAs<QubitTimeMetrics>(qubitsTraceData);
-            opRec.MaxOperationStartTime = QubitsMetricsUtils.MaxQubitAvailableTime(opRec.InputQubitMetrics);
-            opRec.MinOperationStartTime = QubitsMetricsUtils.MinQubitAvailableTime(opRec.InputQubitMetrics);
+            opRec.MaxOperationStartTime = MaxAvailableTime(opRec.InputQubitMetrics);
+            opRec.MinOperationStartTime = MinAvailableTime(opRec.InputQubitMetrics);
+            opRec.MaxQubitIdAtStart = qubitAvailabilityTime.GetMaxQubitId();
             operationCallStack.Push(opRec);
         }
 
+        private double MinAvailableTime(QubitTimeMetrics[] qubitTimeMetrics)
+        {
+            Debug.Assert(qubitTimeMetrics != null);
+            double min = Double.MaxValue;
+            foreach (QubitTimeMetrics metric in qubitTimeMetrics)
+            {
+                min = Min(min, qubitAvailabilityTime[metric.QubitId]);
+            }
+            return min != Double.MaxValue ? min : 0;
+        }
+
+        private double MaxAvailableTime(QubitTimeMetrics[] qubitTimeMetrics)
+        {
+            Debug.Assert(qubitTimeMetrics != null);
+            double max = 0;
+            foreach (QubitTimeMetrics metric in qubitTimeMetrics)
+            {
+                max = Max(max, qubitAvailabilityTime[metric.QubitId]);
+            }
+            return max;
+        }
```

Comment:
foreach makes it clear what's going on here instead of obscure methods and lambda function to compute per element. I will consider using enumerable though to avoid some array materialization in callers.

Reply:
I'm not sure I'd refer to core C# language features as obscure, but fair enough. Mostly my comment here was with respect to this was the duplicated code as a result.